src/test/scala/chiselTests/InstanceSpec.scala

```diff
@@ -0,0 +1,444 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chiselTests
+
+import chisel3._
+import chisel3.stage.ChiselStage
+import chisel3.testers.BasicTester
+import chisel3.experimental.annotate
+import chisel3.internal.{interface, instance, public}
+import _root_.firrtl.annotations._
+
+@instance
```

Comment:
Why do you annotate the module with `@instance` when it seems like this is actually the module _definition_ and not the place where we create an instance?

Reply:
this isn't the definition, either (the way we are using "definition" in this PR), i think maybe "@instanceable" is the right flavor of the intention here...

Reply:
We could call it `@definition`?

Reply:
I don't think `@definition` is right. The definition is (as you're using it here) when you call this to make an explicit `Definition` of a thing you can make `Instance`s of.

Reply:
Could we enable this feature for every Module by default? Like is there a way to support making instances without having to annotate the module? That way libraries would not have to migrate. Like e.g., with the current solution it seems like we will have to annotate all the modules in `chisel3.util` to allow people to make instances of them.

Reply:
It has been renamed to `@instantiable`

> Could we enable this feature for every Module by default?

You can create an `Instance` of any `Module` even without `@instantaible`, but you won't have the nice `.` syntax for fields, instead you'll have to use `myInstance(_.field)`. This is fundamental to the approach although I guess we could look into generating the extension methods in the compiler plugin instead of via macros. Similar issue for `@public`.

Reply:
Oh cool. Then maybe we could have the plugin expose all IOs by default.

Reply:
> Then maybe we could have the plugin expose all IOs by default.

I don't think we can unless it's "all vals of type Data" by default because there is no type information known to the Scala compiler about what is an IO vs. what isn't. There are also serious issues with inheritance (that also apply with the macro so I'll call that one a wash)

I also just remembered why Adam used a macro instead of the plugin in the first place which is that whitebox stuff (eg. generating methods that you can then call) is suuuuuper hard to do in plugins and hard enough with macros so we might be about at the limit of what we can do with this approach here.

Reply:
OK. No compiler plugin magic in this PR for sure! I am just hopeful that maybe in the future we will figure out a way.

Reply:
Yeah, my position is that this API is "aggressively experimental." We've chosen a strategy and we're taking it as far as we can to make the best user API we can. It may be that a different approach is better and we should continue to explore.